File: plusOne.js
function plusOne(input){
    let end = input.length-1;
    let i = end;
    input[i]+=1;

    while(i>0 && input[i]===10) {
        input[i]=0;
        i-=1;
        input[i]+=1;
    }

    if(input[0]===10) {
        input[0]=0;
        input.unshift(1);
    }

    return input;
}

console.log(plusOne([1,2,9])) // 1,3,0
console.log(plusOne([1,9,9])) // 2,0,0
console.log(plusOne([9,9,9])) // 1,0,0,0
console.log(plusOne([9])) // 1,0
console.log(plusOne([1,1,1,9])) // 1,1,2,0

File: plusOne.java
  public int[] plusOne(int[] digits) {
      int extra = 1;
      int temp = 0;
      for(int i=digits.length-1; i>=0; i--){
          digits[i] = extra+digits[i];
          temp = digits[i]%10;
          extra = digits[i]/10;
          digits[i] = temp;                
      }
      if(extra > 0){
          int[] result = new int[digits.length+1];
          result[0] = 1;
          System.arraycopy(digits, 0, result, 1, digits.length);
          return result;
      } else {
          return digits;
      }
  }
